src/main/scala/lms/core/backend.scala

```diff
@@ -276,6 +269,26 @@ class GraphBuilder {
     }
   }
 
+  def gatherEffectDeps(reads: Set[Exp], writes: Set[Exp], s: String, as: Def*): (Set[Sym], Set[Sym]) = {
+    val (prevHard, prevSoft) = (new mutable.ListBuffer[Sym], new mutable.ListBuffer[Sym])
+    // gather effect dependencies 1): handle the write keys
+    writes.foreach{key => curEffects.get(key) match {
+      case Some((lw, lr)) =>
+        // write has hard dependencies on previous write (this is conservative for array case, Store, Ctrl,...)
+        prevHard += latest(lw);
+        // write has soft dependencies on previous read (just enforcing order, do not enforcing the reads to be scheduled)
+        if (!reflectHere) prevSoft ++= lr else Nil // FIXME(feiw) why not adding soft dependencies when not reflectHere?
+      case _ =>
+        // write has hard dependencies on declaration (if declared locally) or block (if declared globally, i.e., out of current block)
+        prevHard += latest(key);
+    }}
+    // gather effect dependencies 2): if we enforce reifyHere (reify in current block), add curBlock as hard dependencies
```

Comment:
Thanks a lot @GSAir . Indeed, I do not really understand the reasoning behind `reifyHere` (or the `reflectHere` in this same file). My understanding is that, by adding `curBlock` to the hard deps, it forces the node to be scheduled `within` the world of the `curBlock` (i.e., it cannot be `code-motioned` to out of the block).

Let's investigate a bit more and come up with a more appropriate comment for this piece of code.

Reply:
the "code motion" is not really the point of effects. I know it is part of it, but it is more important to see it as: in order to produce the correct result this statement has to be generated before that statement.

This is also the case with vanilla reify/reflect so that's not the particularity of the *Here methods. I am pretty sure that *Here are just optimization, it should work with the vanilla ones. You said you found code difference without the *Here. Can you show an example? It would explain why it is interesting to have the *Here methods.

Reply:
That difference was a mistake. I turned off `reifyHere` but still used `isPure` to compute the effects of `if` nodes, which is just wrong.... It needs more careful investigation to see what is optimized by `reifyHere`